src/GraphQL.Client.Serializer.SystemTextJson/PolymorphicTypenameConverter.cs

```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace GraphQL.Client.Serializer.SystemTextJson
+{
+    public abstract class PolymorphicTypenameConverter<T> : JsonConverter<T>
+    {
+        protected abstract Type Descriminator(string typename);
+
+        public override bool CanConvert(Type typeToConvert) =>
+            typeof(T).IsAssignableFrom(typeToConvert);
+
+        public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            var clone = reader; // cause its a struct
+
+            if(clone.TokenType == JsonTokenType.StartObject)
+                clone.Read();
+            if(clone.TokenType != JsonTokenType.PropertyName || clone.GetString() != "__typename")

```

Comment:
So it works only if I query `__typename` meta-field.

Reply:
Right. its designed to help with using Graphql Fragments. No other purpose.